### PR TITLE
[7264] Add Exception to Broker Response When Not All Segments Are Available (Partial Response)

### PR DESCRIPTION
**Pull request by suddendust**
## Description

This PR added exception to the broker response in the following cases:
1. `numUnavailableSegments` != 0. Implying there are some offline segments.
2. `numSegmentsQueried > numSegmentsAcquired`. That is, the servers could not acquire some segments. 
3. `numServersQueried > numServersResponded`. That is, some servers did not respond.

## Upgrade Notes
Does this PR prevent a zero down-time upgrade?
* [ ] Yes

Does this PR fix a zero-downtime upgrade introduced earlier?
* [ ] Yes

Does this PR otherwise need attention when creating release notes? Things to consider:
* [ ] Yes 

## Documentation

This would require the docs to be upgraded as there are new exceptions in the response.

### CODE REVIEW

**Comment by suddendust** [comment]
Adding tests now.

**Comment by suddendust** [comment]
@Jackie-Jiang I could not find any test that validates that there are no duplicate query exception error codes. I think it'll be a good idea to validate it with a conformance test. What do you think?

File: pinot-core/src/main/java/org/apache/pinot/core/query/executor/ServerQueryExecutorV1Impl.java [hunk 1]
```diff
@@ -167,10 +167,7 @@ public DataTable processQuery(ServerQueryRequest queryRequest, ExecutorService e
     // TODO: Change broker to watch both IdealState and ExternalView to not query the removed segments
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/query/executor/ServerQueryExecutorV1Impl.java:167]: This `TODO` is already done. You may remove this comment block

File: pinot-core/src/main/java/org/apache/pinot/core/query/executor/ServerQueryExecutorV1Impl.java [hunk 2]
```diff
@@ -251,6 +248,14 @@ public DataTable processQuery(ServerQueryRequest queryRequest, ExecutorService e
           .put(MetadataKey.MIN_CONSUMING_FRESHNESS_TIME_MS.getName(), Long.toString(minConsumingFreshnessTimeMs));
     }
 
+    if (numSegmentsQueried > numSegmentsAcquired) {
+      String errorMessage =
+          String.format("Some segments could not be acquired: %d", numSegmentsQueried - numSegmentsAcquired);
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/query/executor/ServerQueryExecutorV1Impl.java:253]: Collect the segment names unable to be acquired and put them into the error message

File: pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/BaseBrokerRequestHandler.java [hunk 3]
```diff
@@ -794,6 +795,11 @@ private BrokerResponseNative handlePQLRequest(long requestId, String query, Json
       _brokerMetrics.addMeteredTableValue(rawTableName, BrokerMeter.BROKER_RESPONSES_WITH_NUM_GROUPS_LIMIT_REACHED, 1);
     }
 
+    if (0 != numUnavailableSegments) {
```
Comment (Jackie-Jiang) [inline pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/BaseBrokerRequestHandler.java:798]: Let's collect the unavailable segment names and put them into the error message

File: pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/SingleConnectionBrokerRequestHandler.java [hunk 4]
```diff
@@ -126,6 +126,9 @@ protected BrokerResponseNative processBrokerRequest(long requestId, BrokerReques
       _brokerMetrics.addMeteredTableValue(rawTableName, BrokerMeter.BROKER_RESPONSES_WITH_PROCESSING_EXCEPTIONS, 1);
     }
     if (numServersQueried > numServersResponded) {
+      String errorMessage = String.format("Some servers did not respond: %d", numServersQueried - numServersResponded);
```
Comment (Jackie-Jiang) [inline pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/SingleConnectionBrokerRequestHandler.java:129]: Let's collect the missing server name from the `ServerRoutingInstance`

File: pinot-common/src/main/java/org/apache/pinot/common/exception/QueryException.java [hunk 5]
```diff
@@ -63,19 +63,27 @@ public static void setMaxLinesOfStackTrace(int maxLinesOfStackTrace) {
   public static final int COMBINE_GROUP_BY_EXCEPTION_ERROR_CODE = 600;
   public static final int QUERY_VALIDATION_ERROR_CODE = 700;
   public static final int UNKNOWN_ERROR_CODE = 1000;
+  public static final int SEGMENTS_NOT_ONLINE_ERROR_CODE = 1001;
+  public static final int SEGMENTS_NOT_ACQUIRED_ERROR_CODE = 1002;
+  public static final int SERVERS_NO_RESPONSE_ERROR_CODE = 1003;
```
Comment (Jackie-Jiang) [inline pinot-common/src/main/java/org/apache/pinot/common/exception/QueryException.java:68]:
```suggestion
  public static final int SERVER_NOT_RESPONDING_ERROR_CODE = 270;
```

File: pinot-common/src/main/java/org/apache/pinot/common/exception/QueryException.java [hunk 6]
```diff
@@ -63,19 +63,27 @@ public static void setMaxLinesOfStackTrace(int maxLinesOfStackTrace) {
   public static final int COMBINE_GROUP_BY_EXCEPTION_ERROR_CODE = 600;
   public static final int QUERY_VALIDATION_ERROR_CODE = 700;
   public static final int UNKNOWN_ERROR_CODE = 1000;
+  public static final int SEGMENTS_NOT_ONLINE_ERROR_CODE = 1001;
+  public static final int SEGMENTS_NOT_ACQUIRED_ERROR_CODE = 1002;
```
Comment (Jackie-Jiang) [inline pinot-common/src/main/java/org/apache/pinot/common/exception/QueryException.java:67]:
```suggestion
  public static final int SERVER_SEGMENT_MISSING_ERROR_CODE = 235;
```

File: pinot-common/src/main/java/org/apache/pinot/common/exception/QueryException.java [hunk 7]
```diff
@@ -63,19 +63,27 @@ public static void setMaxLinesOfStackTrace(int maxLinesOfStackTrace) {
   public static final int COMBINE_GROUP_BY_EXCEPTION_ERROR_CODE = 600;
   public static final int QUERY_VALIDATION_ERROR_CODE = 700;
   public static final int UNKNOWN_ERROR_CODE = 1000;
+  public static final int SEGMENTS_NOT_ONLINE_ERROR_CODE = 1001;
```
[the remark this answers is not in the document]
Comment (Jackie-Jiang) [reply pinot-common/src/main/java/org/apache/pinot/common/exception/QueryException.java:66]:
```suggestion
  public static final int BROKER_SEGMENT_UNAVAILABLE_ERROR_CODE = 310;
```

**Review by Jackie-Jiang — COMMENTED** [review]
I feel it might be hard to add a test to avoid adding duplicate error code in the future. IMO having tests on the expected error code is good enough

File: pinot-common/src/main/java/org/apache/pinot/common/exception/QueryException.java (hunk 6 above)
Comment (suddendust) [reply pinot-common/src/main/java/org/apache/pinot/common/exception/QueryException.java:67]: Hey @Jackie-Jiang may are these error codes suggested by you random or is there some logic to this?

Comment (Jackie-Jiang) [reply pinot-common/src/main/java/org/apache/pinot/common/exception/QueryException.java:67]: They are ordered in query execution phases. E.g currently error code 200-250 is server error; 300+ is broker error

**Comment by suddendust** [comment]
@Jackie-Jiang I need some help with setting up mocks for testing these changes on the broker. I went through the code and I see there are no tests for non-literal queries on the broker (where the broker actually does a scatter-gather to/from the servers). Is there a reference where the downstream servers are mocked? Thanks!

File: pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/BaseBrokerRequestHandler.java [hunk 8]
```diff
@@ -487,6 +491,11 @@ private BrokerResponseNative handleSQLRequest(long requestId, String query, Json
...
```
Comment (Jackie-Jiang) [inline pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/BaseBrokerRequestHandler.java:494]: （code convention) We usually put `numUnavailableSegments != 0`. Suggest keeping it consistent for readability. Same for other places

File: pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/BaseBrokerRequestHandler.java [hunk 9]
```diff
@@ -496,6 +505,7 @@ private BrokerResponseNative handleSQLRequest(long requestId, String query, Json
     logBrokerResponse(requestId, query, requestStatistics, brokerRequest, numUnavailableSegments, serverStats,
         brokerResponse, totalTimeMs);
     return brokerResponse;
+
```
Comment (Jackie-Jiang) [inline pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/BaseBrokerRequestHandler.java:508]: (nit) extra empty line

File: pinot-common/src/main/java/org/apache/pinot/common/exception/QueryException.java [hunk 10]
```diff
@@ -63,19 +63,27 @@ public static void setMaxLinesOfStackTrace(int maxLinesOfStackTrace) {
   public static final int COMBINE_GROUP_BY_EXCEPTION_ERROR_CODE = 600;
   public static final int QUERY_VALIDATION_ERROR_CODE = 700;
   public static final int UNKNOWN_ERROR_CODE = 1000;
+  public static final int BROKER_SEGMENT_UNAVAILABLE_ERROR_CODE = 310;
```
Comment (Jackie-Jiang) [inline pinot-common/src/main/java/org/apache/pinot/common/exception/QueryException.java:66]: Order all error code in ascending order

**Comment by Jackie-Jiang** [comment]
> @Jackie-Jiang I need some help with setting up mocks for writing UTs for these changes on the broker. I went through the code and I see there are no tests for non-literal queries on the broker (where the broker actually does a scatter-gather to/from the servers). Is there a reference where the downstream servers are mocked? That'd of great help. Thanks!

There are fake server instance in `ControllerTestUtils`, but not sure if it is possible to reuse it to test the error behavior. Another option is to test it in an integration test

File: pinot-core/src/main/java/org/apache/pinot/core/query/executor/ServerQueryExecutorV1Impl.java [hunk 11]
```diff
@@ -164,13 +166,9 @@ public DataTable processQuery(ServerQueryRequest queryRequest, ExecutorService e
     // 4. Broker watches ExternalView change and updates the routing table to stop querying the segment
     //
     // After step 2 but before step 4, segment will be missing on server side
-    // TODO: Change broker to watch both IdealState and ExternalView to not query the removed segments
```
Comment (jtao15) [inline pinot-core/src/main/java/org/apache/pinot/core/query/executor/ServerQueryExecutorV1Impl.java:167]: Good catch!

File: pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/BaseBrokerRequestHandler.java (hunk 8 above)
Comment (mqliang) [inline pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/BaseBrokerRequestHandler.java:494]: variable `numUnavailableSegments` seems redundant, use 
* `requestStatistics.setNumUnavailableSegments(unavailableSegments.size())` at L445
* `if (! unavailableSegments.isEmpty())` at L494
Same for other places

File: pinot-core/src/test/java/org/apache/pinot/core/query/executor/QueryExecutorExceptionsTest.java [hunk 12]
```diff
@@ -0,0 +1,153 @@
...
+
+public class QueryExecutorExceptionsTest {
+  private static final String AVRO_DATA_PATH = "data/simpleData200001.avro";
+  private static final String EMPTY_JSON_DATA_PATH = "data/test_empty_data.json";
+  private static final String QUERY_EXECUTOR_CONFIG_PATH = "conf/query-executor.properties";
+  private static final File INDEX_DIR = new File(FileUtils.getTempDirectory(), "QueryExecutorTest");
+  private static final String TABLE_NAME = "testTable";
+  private static final int NUM_SEGMENTS_TO_GENERATE = 2;
+  private static final int NUM_EMPTY_SEGMENTS_TO_GENERATE = 2;
+  private static final Pql2Compiler COMPILER = new Pql2Compiler();
```
Comment (mqliang) [inline pinot-core/src/test/java/org/apache/pinot/core/query/executor/QueryExecutorExceptionsTest.java:57]: We have deprecated PQL. All PQL related code will be cleaned up in the future. Please try to use SqlComplier

File: pinot-core/src/test/java/org/apache/pinot/core/query/executor/QueryExecutorExceptionsTest.java [hunk 13]
```diff
@@ -0,0 +1,153 @@
...
+  }
+
+  @Test
+  public void testServerSegmentMissingErrorCode() {
+    String query = "SELECT COUNT(*) FROM " + TABLE_NAME;
+    InstanceRequest instanceRequest = new InstanceRequest(0L, COMPILER.compileToBrokerRequest(query));
+    instanceRequest.setSearchSegments(_segmentNames);
+    DataTable instanceResponse = _queryExecutor.processQuery(getQueryRequest(instanceRequest), QUERY_RUNNERS);
+    Map<Integer, String> exceptions = instanceResponse.getExceptions();
+    Assert.assertTrue(exceptions.containsKey(QueryException.SERVER_SEGMENT_MISSING_ERROR_CODE));
```
Comment (mqliang) [inline pinot-core/src/test/java/org/apache/pinot/core/query/executor/QueryExecutorExceptionsTest.java:139]: better to check the missing segment name list also.

Comment (suddendust) [reply pinot-core/src/test/java/org/apache/pinot/core/query/executor/QueryExecutorExceptionsTest.java:147]: Asserted on the name list.

File: pinot-core/src/main/java/org/apache/pinot/core/query/executor/ServerQueryExecutorV1Impl.java (hunk 1 above)
Comment (suddendust) [reply pinot-core/src/main/java/org/apache/pinot/core/query/executor/ServerQueryExecutorV1Impl.java]: Removed todo.

File: pinot-core/src/main/java/org/apache/pinot/core/query/executor/ServerQueryExecutorV1Impl.java (hunk 2 above)
Comment (suddendust) [reply pinot-core/src/main/java/org/apache/pinot/core/query/executor/ServerQueryExecutorV1Impl.java]: Addressed.

File: pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/BaseBrokerRequestHandler.java (hunk 3 above)
Comment (suddendust) [reply pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/BaseBrokerRequestHandler.java:810]: Addressed.

File: pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/BaseBrokerRequestHandler.java (hunk 8 above)
Comment (suddendust) [reply pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/BaseBrokerRequestHandler.java:494]: Done.

File: pinot-core/src/test/java/org/apache/pinot/core/query/executor/QueryExecutorExceptionsTest.java (hunk 12 above)
Comment (suddendust) [reply pinot-core/src/test/java/org/apache/pinot/core/query/executor/QueryExecutorExceptionsTest.java]: Done

File: pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/BaseBrokerRequestHandler.java (hunk 8 above)
Comment (suddendust) [reply pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/BaseBrokerRequestHandler.java]: Removed `numUnavailableSegments`

**Review by Jackie-Jiang — APPROVED** [review]
LGTM, thanks for the contribution!

**Pull request merged**